packages/kbn-alerts-ui-shared/src/common/types/tanstack_query_utility_types.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { UseQueryOptions } from '@tanstack/react-query';
+
+/**
+ * Extracts the data type from a fetching function
+ */
+export type FetchFnData<TFetchFn> = TFetchFn extends (...args: any) => Promise<infer TData>
+  ? TData
+  : TFetchFn extends (...args: any) => infer TData
+  ? TData
+  : never;
+
+/**
+ * A utility type used to correctly type options overrides for `useQuery`
+ */
+export type QueryOptionsOverrides<
```

Comment:
Do we really need another complicated derived type for this? haha I feel like we should try to keep these simple so it's more readable.

Reply:
IMO it's fine to be verbose and have some repetition to avoid layers and layers of abstraction + redirection just to figure out what types it needs.

Reply:
I actually started by defining the type of each overridable option individually but ended up creating a utility type because some options have very complicated types that, when specified manually, can easily cause tricky type errors at the `useQuery` level, when the overrides are applied.
It you think the added complexity is not worth it then I'll remove these, let's just consider that specifying the types of some options manually might not be always easy

Reply:
What would it look like if we we're to specify the types individually?

Reply:
This is based on the option overrides that are actually used for the `useFetchAlertsFieldsQuery` hook:

```ts
options?: {
    context?: React.Context<QueryClient | undefined>;
    // This should be TError but we don't have access to the generic here...
    onError?: (error: Error) => void;
    refetchOnWindowFocus?:
      | boolean
      | 'always'
      // To type this correctly we need to know all the generic types
      | ((query: Query<TQueryFnData, TError, TQueryData, TQueryKey>) => boolean | 'always');
    staleTime?: number;
    enabled?: boolean;
  }
```

Idk as soon as you want to have full compatibility with the original types it feels a bit too complex

Reply:
> IMO it's fine to be verbose and have some repetition to avoid layers and layers of abstraction + redirection just to figure out what types it needs.

In general, I am always in favor of this philosophy. In this case, though, it is justifiable to use a TS utility to abstract the complexities of the types of React Query. I think the benefits it offers outweigh the downside of the complexity. With this TS type, we can:

- Enable/disable options to all hooks at once
- Make it easier for developers to type their hooks.
- Abstracts the complexities of React Querie's types.
- In case of a breaking change (upgrade to v5) we would need to change one type instead of multiple.

Reply:
I much rather we do something like 

```
interface options {
  refetchOnWindowFocus?: QueryOptions['refetchOnWindowFocus'];
  onError?: QueryOptions['onError'];
  ...
} 
```

And if we are finding that it's becoming extremely verbose then I think it's trying to tell us that maybe we are doing something we shouldn't be doing (like redefining every single useQuery option). If that's the case maybe consider: 

```
interface options {
  myOption: Something;
  myOtherOption: Something;
  useQueryOptions: QueryOptions; <<<<< Group all of the useQueryOptions together
}
```

In general I think we should avoid mixing our own types with existing types (like spreading a bunch of options into a react component). This way it is very clear what properties this type is looking for by looking it and we won't be mixing name spaces.